Adding transactions (#260)
SINTEF/dlite: doc/transactions.md

Transactions
============

> Transactions is a feature that comes from SOFT, which allows to easy manage arbitrary long series of immutable (frozen) instances while ensuring provenance.  Conceptually it share many similarities with git.

Reviewer: ```suggestion
Transactions is a feature that comes from SOFT, which allows to easy manage arbitrary long series of immutable (frozen) instances, while ensuring provenance.
Conceptually, it shares many similarities with [git](https://git-scm.com).
```

> A basic usage for transactions is to take snapshots of the current state of your system.  This is shown in Figure 1, where we assume that the instance `A` (which e.g. could be a collection) describes the state of your system.

Reviewer: ```suggestion
A basic usage for transactions is to take snapshots of the current state of your system.
This is shown in Figure 1, where we assume that the instance `A` (which, e.g., could be a collection) describes the state of your system.
```

> We call an instance a *transaction* after we have taken a snapshot of it, i.e. when it contains a reference to a immutable "parent" instance.

Reviewer: ```suggestion
We call an instance a *transaction* after we have taken a snapshot of it, i.e. when it contains a reference to an immutable previous ("parent") instance.
```

![transactions](figs/transactions.png)

> **Figure 1**: *Creating a transaction by taking snapshots.  (a) initial state of your system described by instance `A`.  (b) creating a transaction by taking a snapshot
> of `A`.  The snapshot, `A1`, stores an immutable (frozen) copy of `A` from the exact
> moment `t1` the snapshot was taken.  (c) After another snapshot was taken at time `t2`.  Blue circles represent immutable instances, red circles represent mutable instances, while arrows relate an instance to its (frozen) parent instance.*

Reviewer: ```suggestion
**Figure 1**: *Creating a transaction by taking snapshots.
(a) initial state of your system described by instance `A`.
(b) creating a transaction by taking a snapshot of `A`.
The snapshot, `A1`, stores an immutable (frozen) copy of `A` from the exact moment `t1` the snapshot was taken.
(c) After another snapshot was taken at time `t2`.
Blue circles represent immutable instances, red circles represent mutable instances, while arrows relate an instance to its (frozen) parent instance.*
```

> It is very simple to create such snapshots using the Python and C APIs.  In Python, the steps in Figure 1 can be produced by the following code:

Reviewer: ```suggestion
It is straight forward to create such snapshots using the Python and C APIs.
In Python, the steps in Figure 1 can be produced by the following code:
```

```python
A = dlite.get_instance("A")  # Initial state (Fig. 1a)
A.snapshot()                 # Make snapshot at time=t1 (Fig. 1b)
# A is evolved...
A.snapshot()                 # Make snapshot at time=t2 (Fig. 1c)
```

The corresponding code in C would be:

```C
DLiteInstance *A = dlite_instance_get("A");  // Initial state (Fig. 1a)
dlite_instance_snapshot(A);                  // Make snapshot at time=t1 (Fig. 1b)
// A is evolved...
dlite_instance_snapshot(A);                  // Make snapshot at time=t2 (Fig. 1c)
```

> The snapshots can be accessed using `dlite_instance_get_snapshot()`.  For instance, accessing snapshot `A2` and creating a new branch from it can be achieved by the following four lines of Python C code:

Reviewer: ```suggestion
The snapshots can be accessed using `dlite_instance_get_snapshot()`.
For instance, accessing snapshot `A2` and creating a new branch from it can be achieved by the following four lines of Python code:
```

```python

> A2 = A.get_snapshot()  # Access the most resent snapshot of A

Reviewer: ```suggestion
A2 = A.get_snapshot()  # Access the most recent snapshot of A
```

B = A2.copy()          # Create a mutable copy of A2

> B.set_parent(A2)       # Make B a transaction with A2 as parent. Shown Fig. 2a.

Reviewer: ```suggestion
B.set_parent(A2)       # Make B a transaction with A2 as parent. Shown in Fig. 2a.
```

> B.shapshot()           # Make a snapshop of B. Shown Fig. 2b.

Reviewer: ```suggestion
B.shapshot()           # Make a snapshop of B. Shown in Fig. 2b.
```

```

The corresponding C code is:

```C

> // Access the most resent snapshot of instance A.

Reviewer: ```suggestion
// Access the most recent snapshot of instance A.
```

> // Note that A2 is a borrowed reference to the snapshot and should not be dereferred

Reviewer: "dereferred" or "dereferenced"?

Author: dereferenced is better

// with dlite_instance_decref().
const DLiteInstance *A2 = dlite_instance_get_snapshot(A, 1);

// Create a mutable copy of A2.
// Note that B is a new reference.  You should call dlite_instance_decref() when you
// are done with it.
DLiteInstance *B = dlite_instance_copy(A2);

> // Make B a transaction with A2 as parent. Shown Fig. 2a.

Reviewer: ```suggestion
// Make B a transaction with A2 as parent. Shown in Fig. 2a.
```

dlite_instance_set_parent(B, A2);

> // Make a snapshop of B. Shown Fig. 2b.

Reviewer: ```suggestion
// Make a snapshop of B. Shown in Fig. 2b.
```

dlite_instance_snapshot(B);
```

> The result if these commands are shown in Figure 2.

Reviewer: ```suggestion

The result if these commands are shown in Figure 2.
```

![transactions-branch](figs/transactions-branch.png)

> **Figure 2**: *Creating a new branch, from a snapshot of a transaction.  (a) create a copy `B` of latest snapshot of transaction `A`.  (b)  Take a snapshot of `B`.*

Reviewer: ```suggestion
**Figure 2**: *Creating a new branch, from a snapshot of a transaction.
(a) Create a copy (`B`) of latest snapshot of transaction `A`.
(b) Take a snapshot of `B`.*
```

> 

Reviewer: ```suggestion
```

### Transaction parent and immutability

> All transactions starts with a root instance with no parent instance.  All other instances in a transaction has exactly one parent instance.

Reviewer: ```suggestion

All transactions start as a root instance with no parent instance.
All other instances in a transaction has exactly **one** parent instance.
```

> All instances in a transaction that serves as a parent are immutable (that is, all instances except the leafs).  Non-root transaction instances store a [SHA-3](https://en.wikipedia.org/wiki/SHA-3) hash of their parents together with the parent UUIDs.  This make it possible to ensure that any of the ancestors of a transaction has not been changed - providing provenance.

Reviewer: ```suggestion
All instances in a transaction that serves as a parent are immutable (that is, all instances except the leaves).
Non-root transaction instances store a [SHA-3](https://en.wikipedia.org/wiki/SHA-3) hash of their parent together with the parent UUID.
This make it possible to ensure that any of the ancestors of a transaction has not been changed - providing provenance.
```

A transaction can be validated with the `verify_transaction()` method in Python and `dlite_instance_verify_transaction()` in C.

> 
> ### Memory management

Reviewer: ```suggestion
### Memory management

```

> Since the number of snapshots potentially can be very large, it is important to be able to store them to disk in order to save memory.  To support this, DLite implements the `pull_snapshot()` and `push_snapshot()` methods (`dlite_instance_pull_snapshot()` and `dlite_instance_push_snapshot()` in C).

Reviewer: ```suggestion
The number of snapshots can potentially be very large, hence it is important to be able to store them to disk in order to save memory.
To support this, DLite implements the `pull_snapshot()` and `push_snapshot()` methods (`dlite_instance_pull_snapshot()` and `dlite_instance_push_snapshot()` in C).
```

Transactions

> ============
> Transactions is a feature that comes from SOFT, which allows to easy manage arbitrary long series of immutable (frozen) instances, while ensuring provenance.

Reviewer: ```suggestion
============

Transactions is a feature that comes from SOFT, which allows to easy manage arbitrary long series of immutable (frozen) instances, while ensuring provenance.
```

Reviewer: Do you have a reference to the part in SOFT where this comes from? If not, that's fine.

Conceptually, it shares many similarities with [git](https://git-scm.com).

A basic usage for transactions is to take snapshots of the current state of your system.
This is shown in Figure 1, where we assume that the instance `A` (which, e.g., could be a collection) describes the state of your system.

> We call an instance a *transaction* after we have taken a snapshot of it, i.e. when it contains a reference to an immutable previous ("parent") instance.

Reviewer: Is it possible to take a snapshot that is exactly similar in content to a previous one? This should be mentioned somewhere in this section.

In the figure example, e.g., it is unclear whether the _values_/content of `A1` has been changed, or the time `t1` is merely a meta-data value.
In general, the concept of time in these examples are a bit tenuous at best. It should be made more explicit - either by setting up a thought example about a time-resolved experiment, where time is a value and property of the underlying data model, or in another way that makes it clear what `t1` and `t2` are, other than abstract moments in time.

Reviewer 2: Just to be clear about nomenclature. When dealing with transactions we have to separate between identity, state and values. A value cannot change, 3 will always be 3, but the age of an item can change over time, but at any given time have a specific state. The state at a given point in time (specific instance of state/snapshot) is a value, and cannot be changed. identity is related to a sequence of states (snapshots/transactions) and point to different values (immutable) at a specific point in time (state).

B.shapshot()           # Make a snapshop of B. Shown in Fig. 2b.
```

The corresponding C code is:

```C
// Access the most recent snapshot of instance A.
// Note that A2 is a borrowed reference to the snapshot and should not be
// dereferenced with dlite_instance_decref().
const DLiteInstance *A2 = dlite_instance_get_snapshot(A, 1);

// Create a mutable copy of A2.
// Note that B is a new reference.  You should call dlite_instance_decref() when you
// are done with it.
DLiteInstance *B = dlite_instance_copy(A2);

// Make B a transaction with A2 as parent. Shown in Fig. 2a.
dlite_instance_set_parent(B, A2);

// Make a snapshop of B. Shown in Fig. 2b.
dlite_instance_snapshot(B);
```
The result of these commands are shown in Figure 2.

![transactions-branch](figs/transactions-branch.png)

**Figure 2**: *Creating a new branch, from a snapshot of a transaction.
(a) Create a copy (`B`) of latest snapshot of transaction `A`.
(b) Take a snapshot of `B`.*

### Transaction parent and immutability

All transactions start as a root instance with no parent instance.
All other instances in a transaction has exactly **one** parent instance.

> All instances in a transaction that serves as a parent are immutable (that is, all instances except the leaves).

Reviewer: ```suggestion
All instances in a transaction that serve as a parent are immutable (that is, all instances except the leaves (the latest)).
```

Non-root transaction instances store a [SHA-3](https://en.wikipedia.org/wiki/SHA-3) hash of their parent together with the parent UUID.

> This make it possible to ensure that any of the ancestors of a transaction has not been changed - providing provenance.

Reviewer: ```suggestion
This makes it possible to ensure that any of the ancestors of a transaction have not been changed - providing provenance.
```

> A transaction can be validated with the `verify_transaction()` method in Python and `dlite_instance_verify_transaction()` in C.

Reviewer: Since the method names use "verify", it seemed more appropriate to use in the text as well, I think.
```suggestion
A transaction can be verified with the `verify_transaction()` method in Python and `dlite_instance_verify_transaction()` in C.
```

### Memory management

The number of snapshots can potentially be very large, hence it is important to be able to store them to disk in order to save memory.
To support this, DLite implements the `pull_snapshot()` and `push_snapshot()` methods (`dlite_instance_pull_snapshot()` and `dlite_instance_push_snapshot()` in C).

> Note that not all storages can be used with these functions, since whether an instance is a transactions or not, is not described by its metadata and hence requires special support by the storage plugin.

Reviewer: It might be nice to write a bit more about the future of memory management, specifically how it's stored on disk to save space and such? It doesn't need to be much, just some empty promises that similar to `git`, we might try to pack/bundle old snapshots together to save even more space on disk?

Transactions

> ============

Reviewer: ```suggestion
============

```
